packages/video_player/tizen/src/video_player.cc

```diff
@@ -423,13 +450,83 @@ void VideoPlayer::OnError(int code, void *data) {
 
 void VideoPlayer::OnVideoFrameDecoded(media_packet_h packet, void *data) {
   auto *player = reinterpret_cast<VideoPlayer *>(data);
-
   std::lock_guard<std::mutex> lock(player->mutex_);
-  if (player->current_media_packet_) {
-    LOG_INFO("[VideoPlayer] Media packet already pending.");
-    media_packet_destroy(packet);
+  player->SendMessage(kMessageOnFrameDecoded, packet);
+}
+
+void VideoPlayer::RunMediaPacketLoop(void *data, Ecore_Thread *thread) {
+  auto *self = reinterpret_cast<VideoPlayer *>(data);
+  Eina_Thread_Queue *packet_thread_queue = eina_thread_queue_new();
+  if (!packet_thread_queue) {
+    LOG_ERROR("Invalid packet queue.");
+    ecore_thread_cancel(thread);
     return;
   }
-  player->current_media_packet_ = packet;
-  player->texture_registrar_->MarkTextureFrameAvailable(player->texture_id_);
+
+  self->packet_thread_queue_ = packet_thread_queue;
+  std::queue<media_packet_h> packet_queue;
+  while (!ecore_thread_check(thread)) {
+    void *ref;
+    Message *message = static_cast<Message *>(
+        eina_thread_queue_wait(packet_thread_queue, &ref));
+    if (message->event == kMessageQuit) {
+      eina_thread_queue_wait_done(packet_thread_queue, ref);
+      break;
+    }
+
+    if (message->event == kMessageOnFrameDecoded) {
+      if (message->media_packet != nullptr) {
+        packet_queue.push(message->media_packet);
+      }
+    }
+
+    eina_thread_queue_wait_done(packet_thread_queue, ref);
+    std::lock_guard<std::mutex> lock(self->mutex_);
+    if (packet_queue.empty() || self->is_rendering_) {
+      continue;
+    }
+    if (self->texture_registrar_->MarkTextureFrameAvailable(
+            self->texture_id_)) {
+      self->is_rendering_ = true;
+      self->previous_media_packet_ = self->current_media_packet_;
+      self->current_media_packet_ = packet_queue.front();
+      packet_queue.pop();
+    }
+  }
+
+  while (!packet_queue.empty()) {
+    media_packet_destroy(packet_queue.front());
+    packet_queue.pop();
+  }
+
+  if (packet_thread_queue) {
+    eina_thread_queue_free(packet_thread_queue);
+  }
+}
+
+void VideoPlayer::SendMessage(int event, media_packet_h media_packet) {
+  if (!packet_thread_ || ecore_thread_check(packet_thread_)) {
+    LOG_ERROR("Invalid packet thread.");
+    return;
+  }
+
+  if (!packet_thread_queue_) {
+    LOG_ERROR("Invalid packet thread queue.");
+    return;
+  }
+
+  void *ref;
+  Message *message = static_cast<Message *>(
+      eina_thread_queue_send(packet_thread_queue_, sizeof(Message), &ref));
+  message->event = event;
+  message->media_packet = media_packet;
+  eina_thread_queue_send_done(packet_thread_queue_, ref);
+}
+
+void VideoPlayer::SendRenderFinishedMessage() {
+  if (previous_media_packet_) {
+    media_packet_destroy(previous_media_packet_);
+    previous_media_packet_ = nullptr;
+  }
+  SendMessage(kMessageOnRenderFinished, nullptr);
```

Comment:
What is the `kMessageOnRenderFinished` message for?

Reply:
sent this message to request render if message queue not empty.

Reply:
Hmm, in that case "OnRenderFinished" and "request rendering" do not match in their meaning. "OnRenderFinished" might need better naming.

Also is there any real world case that the packet queue is not empty when the "OnRenderFinished" message is sent? As far as I can see a media packet is queued when a new frame is decoded by the player and then it's immediately sent to the engine for rendering, so the queue should remain empty in most cases. Even if the `lock` is acquired again by `OnVideoFrameDecoded` before the thread acquires the `lock`, the frames will still be rendered in order and the queue will be emptied anyway.

Reply:
> Hmm, in that case "OnRenderFinished" and "request rendering" do not match in their meaning. "OnRenderFinished" might need better naming.
> 
> Also is there any real world case that the packet queue is not empty when the "OnRenderFinished" message is sent? As far as I can see a media packet is queued when a new frame is decoded by the player and then it's immediately sent to the engine for rendering, so the queue should remain empty in most cases. Even if the `lock` is acquired again by `OnVideoFrameDecoded` before the thread acquires the `lock`, the frames will still be rendered in order and the queue will be emptied anyway.

Also is there any real world case that the packet queue is not empty when the "OnRenderFinished" message is sent?
==>when decoding speed is faster than rendering speed.

Reply:
I was not talking about the "On" prefix in the name. I was asking why the rendering "finished" message would "trigger" rendering.

> when decoding speed is faster than rendering speed.

Okay, but this still doesn't explain why the `kMessageRenderFinished` message is used for scheduling a single frame. Let's say the decoding speed is faster than the rendering speed and some `kMessageFrameDecoded` messages have been ignored due to the `is_rendering_` check. If a `kMessageRenderFinished` message is received by the thread in this situation, a packet will be dequeued from `packet_queue`, but even after that `packet_queue` will still remain non-empty if there were more than one packets in the queue. Is this what you expected? Please take more time to explain your code so that it can be understood by average readers.

Reply:
About render queue, we have three types event:
1. FrameDecoded event
  when receive decoded callback we will push the packet to packet queue.
```cpp
    if (message->event == kMessageFrameDecoded) {
      if (message->media_packet != nullptr) {
        packet_queue.push(message->media_packet);
      }
    }
```
  If there is no rendering operation at this time， we can request to render.
```cpp
    if (packet_queue.empty() || self->is_rendering_) {
      continue;
    }
    if (self->texture_registrar_->MarkTextureFrameAvailable(
            self->texture_id_)) {
      self->is_rendering_ = true;
      self->previous_media_packet_ = self->current_media_packet_;
      self->current_media_packet_ = packet_queue.front();
      packet_queue.pop();
    }
```
2. kMessageRenderFinished
we may receive frame decoded callback during rendering operation, we just push the packet to the packet queue, but not requet render. so after rendering, we can can send RenderFinished event to message loop, if  packet queue not empty, we need request to render.
```cpp
    if (packet_queue.empty() || self->is_rendering_) {
      continue;
    }
    if (self->texture_registrar_->MarkTextureFrameAvailable(
            self->texture_id_)) {
      self->is_rendering_ = true;
      self->previous_media_packet_ = self->current_media_packet_;
      self->current_media_packet_ = packet_queue.front();
      packet_queue.pop();
    }
```
3. kMessageQuit
destory message loop

Reply:
Thank you for the explanation. So the render finished event is sent from the `ReleaseMediaPacket` callback to request rendering of the next packet in the queue (if any). In that case the name `SendRenderFinishedMessage` makes sense when used within the release callback but not in `ObtainGpuBuffer` because no rendering actually happened. This is the part that confused me.

Another concern is that if the decoding speed can be faster than the rendering speed, the queue may grow more and more over time and the app may encounter its memory limit. Is that possible? (It will also slow down the video playback and `GetPosition` will report a wrong value as a consequence.) If the packets must be destroyed in decoding order, how about destroying all stale packets in the queue (except the latest one) when `ReleaseMediaPacket` is called?

Reply:
I asked MM team member, they said for HW decoder:
The number of output buffer is fixed, so decoded buffer will not be come if app keeps too many buffers not to rendered.
They suggest that plugin can control this situation by dropping new buffer decoded callback if too many buffers are queued.

Reply:
> dropping new buffer decoded callback

Isn't this the way `OnVideoFrameDecoded` was implemented before this change?

Reply:
> > dropping new buffer decoded callback
> 
> Isn't this the way `OnVideoFrameDecoded` was implemented before this change?

Yes, but message queue solution has another difference, it doesn't destory current media packet when render finished,
These are MM team member guide me to do so.

Reply:
I don't still clearly understand the solution but wouldn't the result be the same without the new thread (`RunMediaPacketLoop`)? i.e. `packet_queue.push` could be moved to `OnVideoFrameDecoded` and `RunMediaPacketLoop` may become a simple method call (might be named `RequestRendering`). `OnVideoFrameDecoded`, `ObtainGpuSurface`, and `ReleaseMediaPacket` are all called asynchronously and the thread seems to add unnecessary complexity on top of it.

Reply:
> I don't still clearly understand the solution but wouldn't the result be the same without the new thread (`RunMediaPacketLoop`)? i.e. `packet_queue.push` could be moved to `OnVideoFrameDecoded` and `RunMediaPacketLoop` may become a simple method call (might be named `RequestRendering`). `OnVideoFrameDecoded`, `ObtainGpuSurface`, and `ReleaseMediaPacket` are all called asynchronously and the thread seems to add unnecessary complexity on top of it.

Yes, you are right, actually we don't need create a new thread, I try update a new version and test it on RB5 and TV devices.